design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+
+This means that a user who has their repo and virtual chunks in Google Cloud Storage, only needs to instruct VirtualiZarr to create URLs with protocol gcs, as in `gcs://foo/bar.nc` and things will just work in Icechunk, without the need to tweak any configuration.
+
+### Object store client infrastructure
+
+A new trait takes care of the "fetching virtual chunks abstraction":
+
+```rust
+#[async_trait]
+pub trait ChunkFetcher: std::fmt::Debug + private::Sealed + Send + Sync {
+    async fn fetch_chunk(
+        &self,
+        chunk_location: &str,
+        range: &ByteRange,
+        checksum: Option<&Checksum>,
+    ) -> Result<Bytes, VirtualReferenceError>;
+}
+
+```
+
+We will have one implementation for each `ObjectStorePlatform` where we want to "speak" a different protocol or optimize differently. For example, we can have one implementation that uses the Rust native S3 client for `s3` and other one that uses object_store for `LocalFileSystem`.
+
+The `Repository` now owns a `VirtualChunkResolver`. This struct is in charge of hosting the containers, matching them and building the `ChunkFetchers` needed.
+
+```rust
+pub struct VirtualChunkResolver {
+    containers: Vec<VirtualChunkContainer>,
+    credentials: HashMap<ContainerName, ObjectStoreCredentials>,
+    fetchers: RwLock<HashMap<ContainerName, Arc<dyn ChunkFetcher>>>,
+}
+```
+
+Creating a new object store client for each virtual chunk request would obviously not work well. We'll have a cache of clients, one per `VirtualChunkContainer` in use. This is not the most efficient thing, we could share the same client across containers that share the platform for example, but it's good enough and simpler. We can always optimize later.
+
+The cached `ChunkFetchers` are maintained by the `VirtualChunkResolver` in its `fetchers` field. Fetchers are instantiated on demand, on the first chunk that needs them and then cached without invalidation. Careful coding is required to avoid a herd effect when multiple chunks for the same container are requested concurrently (a very common situation)
+
+### Credentials
+
+We haven't discussed the third field in the `VirtualChunkResolver`: `credentials`. This field gets populated when the `Repository` is initialized and it holds the credentials passed by the user for each container. Users don't necessarily need to pass credentials for every container, but if they don't, their requests for those chunks may fail.
+
+The `ContainerName` field in the `VirtualChunkContainer` is how the users refer to them to pass credentials.
+
+Several types of credentials are possible.
+
+```rust
+pub enum ObjectStoreCredentials {
+    FromEnv,
+    Anonymous,
+    Static {
+        access_key_id: String,
+        secret_access_key: String,
+        session_token: Option<String>,
+    },
```

Comment:
These are S3-style credentials. We'll need different structs for S3, GCS, and Azure protocols. This spec gives a good enumeration: https://github.com/unitycatalog/unitycatalog/blob/main/api/Models/TemporaryCredentials.md